src/libraries/System.Private.CoreLib/src/System/SearchValues/ProbabilisticWithAsciiCharSearchValues.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.Wasm;
+using System.Runtime.Intrinsics.X86;
+
+namespace System.Buffers
+{
+    internal sealed class ProbabilisticWithAsciiCharSearchValues<TOptimizations> : SearchValues<char>
+        where TOptimizations : struct, IndexOfAnyAsciiSearcher.IOptimizations
+    {
+        private Vector256<byte> _asciiBitmap;
+        private Vector256<byte> _inverseAsciiBitmap;
+        private ProbabilisticMap _map;
+        private readonly string _values;
+
+        public ProbabilisticWithAsciiCharSearchValues(scoped ReadOnlySpan<char> values)
+        {
+            Debug.Assert(IndexOfAnyAsciiSearcher.IsVectorizationSupported);
+            Debug.Assert(values.ContainsAnyInRange((char)0, (char)127));
+
+            IndexOfAnyAsciiSearcher.ComputeBitmap(values, out _asciiBitmap, out _);
+            _inverseAsciiBitmap = ~_asciiBitmap;
+
+            _values = new string(values);
+            _map = new ProbabilisticMap(_values);
+        }
+
+        internal override char[] GetValues() => _values.ToCharArray();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal override bool ContainsCore(char value) =>
+            ProbabilisticMap.Contains(ref Unsafe.As<ProbabilisticMap, uint>(ref _map), _values, value);
+
+        internal override int IndexOfAny(ReadOnlySpan<char> span)
+        {
+            int offset = 0;
+
+            if (IndexOfAnyAsciiSearcher.IsVectorizationSupported && span.Length >= Vector128<short>.Count)
+            {
+                // We are using IndexOfAnyAsciiSearcher to search for the first ASCII character in the set, or any non-ASCII character.
+                // We do this by inverting the bitmap and using the opposite search function (Negate instead of DontNegate).
+
+                // As we're using the inverse bitmap, we have to make sure to use the correct IOptimizations implementation
+                if ((Ssse3.IsSupported || PackedSimd.IsSupported) && typeof(TOptimizations) == typeof(IndexOfAnyAsciiSearcher.Default))
+                {
+                    offset = IndexOfAnyAsciiSearcher.IndexOfAnyVectorized<IndexOfAnyAsciiSearcher.Negate, IndexOfAnyAsciiSearcher.Ssse3AndWasmHandleZeroInNeedle>(
+                        ref Unsafe.As<char, short>(ref MemoryMarshal.GetReference(span)),
+                        span.Length,
+                        ref _inverseAsciiBitmap);
+                }
+                else
+                {
+                    offset = IndexOfAnyAsciiSearcher.IndexOfAnyVectorized<IndexOfAnyAsciiSearcher.Negate, IndexOfAnyAsciiSearcher.Default>(
```

Comment:
I don't understand why using Default is correct here?

Reply:
If the bitmap we're using contains a 0, we have to use `Ssse3AndWasmHandleZeroInNeedle` when running on X86 and WASM. Everything else should use `Default`.

`TOptimizations` specifies whether `_asciiBitmap` contains a 0. Since we're using the inverse bitmap in this case, we have to use `Ssse3AndWasmHandleZeroInNeedle` iff we're on X86/WASM and `TOptimizations` is `Default` (as that means that the inverse bitmap definitely has a 0).

Reply:
Can we both a) put that in a comment, and b) add some asserts?  The logic makes sense, but it seems like it could easily rot in the future if not careful.

Reply:
Done